text/0000-inline-semantic.md

```diff
@@ -0,0 +1,1010 @@
+note: Run with `RUST_BACKTRACE=1` for a backtrace.
+
+$ ./1 arg1 arg2
+args[1] = arg1
+args[2] = arg2
+thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', 1.rs:6:29
+note: Run with `RUST_BACKTRACE=1` for a backtrace.
+
+$ ./1 arg1 arg2 arg3
+args[1] = arg1
+args[2] = arg2
+args[3] = arg3
+```
+
+Let's say you are unhappy with these built-in functions, e.g. you want to provide an alternative
+error message:
+
+```rust
+// 2.rs
+use std::env::args;
+pub fn my_unwrap<T>(input: Option<T>) -> T {
+    match input {
+        Some(t) => t,
+        None => panic!("nothing to see here, move along"),
+    }
+}
+fn main() {
+    println!("args[1] = {}", my_unwrap(args().nth(1)));
+    println!("args[2] = {}", my_unwrap(args().nth(2)));
+    println!("args[3] = {}", my_unwrap(args().nth(3)));
+}
+```
+
+This trivial implementation, however, will only report the panic happens inside `my_unwrap`. This is
+pretty useless, since it is the caller of `my_unwrap` that have made the wrong assumption!
+
+```text
+$ ./2
+thread 'main' panicked at 'nothing to see here, move along', 2.rs:5:16
+note: Run with `RUST_BACKTRACE=1` for a backtrace.
+
+$ ./2 arg1
+args[1] = arg1
+thread 'main' panicked at 'nothing to see here, move along', 2.rs:5:16
+note: Run with `RUST_BACKTRACE=1` for a backtrace.
+
+$ ./2 arg1 arg2
+args[1] = arg1
+args[2] = arg2
+thread 'main' panicked at 'nothing to see here, move along', 2.rs:5:16
+note: Run with `RUST_BACKTRACE=1` for a backtrace.
+
+$ ./2 arg1 arg2 arg3
+args[1] = arg1
+args[2] = arg2
+args[3] = arg3
+```
+
+The trivial solution would be requiring user to provide `file!()`, `line!()` and `column!()`. A
+slightly more ergonomic solution would be changing `my_unwrap` to a macro, thus these constants can
+be automatically provided.
+
+```rust
+pub fn my_unwrap_at_source_location<T>(input: Option<T>, file: &str, line: u32, column: u32) -> T {
+    match input {
+        Some(t) => t,
+        None => panic!("nothing to see at {}:{}:{}, move along", file, line, column),
+    }
+}
+
+macro_rules! my_unwrap {
+    ($input:expr) => {
+        my_unwrap_at_source_location($input, file!(), line!(), column!())
+    }
+}
+println!("args[1] = {}", my_unwrap!(args().nth(1)));
+//                                ^ tell user to add an `!`.
+...
+```
+
+But what if you have already published the `my_unwrap` crate that has thousands of users, and you
+want to maintain API stability? Before Rust 1.XX, the builtin `unwrap()` has the same problem!
+
+## Adding caller location implicitly
+
+The reason a `my_unwrap!` macro works is because it copy-and-paste the entire content of its macro
+definition everytime it is used.
+
+```rust
+println!("args[1] = {}", my_unwrap!(args().nth(1)));
+println!("args[2] = {}", my_unwrap!(args().nth(2)));
+...
+
+// is equivalent to:
+
+println!("args[1] = {}", my_unwrap(args().nth(1), file!(), line!(), column!()));
+println!("args[1] = {}", my_unwrap(args().nth(2), file!(), line!(), column!()));
```

Comment:
The Location struct was not introduced at this point yet. It is still referring to the `my_unwrap!` macro in the previous section.